WcaOnRails/app/models/team_member.rb

```diff
@@ -13,7 +15,25 @@ def current_member?
   validate :start_date_must_be_earlier_than_end_date
   def start_date_must_be_earlier_than_end_date
     if start_date && end_date && start_date >= end_date
-      errors.add(:start_date, "must be earlier than end_date")
+      errors.add(:start_date, "must be earlier than end date")
+    end
+  end
+
+  validate :membership_periods_cannot_overlap_for_single_user
+  def membership_periods_cannot_overlap_for_single_user
+    if (start_date)
+      new_period = start_date..(end_date || Date::Infinity.new)
+      if self.id
+       team_members = TeamMember.where("user_id = :user_id and team_id = :team_id and id <> :team_member_id", {user_id: self.user_id, team_id: self.team_id, team_member_id: self.id})
+      else
+       team_members = TeamMember.where("user_id = :user_id and team_id = :team_id", {user_id: self.user_id, team_id: self.team_id})
+      end
+      team_members.each do |team_member|
+        existing_period = team_member.start_date..(team_member.end_date || Date::Infinity.new)
+        if new_period.overlaps? existing_period
+          errors.add(:user_id, "must not have overlapping membership dates with the #{self.team.name}.")
+        end
+      end
     end
```

Comment:
I was under assumption that there may be many new team members at once (this was possible before). Nice that this can be simplified now. I'd change it a bit:
```ruby
def membership_periods_cannot_overlap_for_single_user
  if start_date
    self_period = start_date..(end_date || Date::Infinity.new)
    other_members = team.team_members.where(user_id: user_id).where.not(id: id)
  
    other_members.each do |team_member|
      other_period = team_member.start_date..(team_member.end_date || Date::Infinity.new)
      if self_period.overlaps? other_period
          errors.add(:user_id, "must not have overlapping membership dates within #{self.team.name}.")
      end
    end
  end
end
```
Note that even if the id is nil this evaluates to `where.not(id: nil)` which should work.